vault/core.go

```diff
@@ -2946,3 +2964,102 @@ type LicenseState struct {
+	ReqPath          string    `json:"request_path"`
+	Method           string    `json:"request_method"`
+	ClientID         string    `json:"client_id"`
+}
+
+func (c *Core) StoreInFlightReqData(reqID string, data *InFlightReqData) {
+	c.inFlightReqData.InFlightReqMap.Store(reqID, data)
+	c.inFlightReqData.InFlightReqCount.Inc()
+}
+
+// FinalizeInFlightReqData is going log the completed request if the
+// corresponding server config option is enabled. It also removes the
+// request from the inFlightReqMap and decrement the number of in-flight
+// requests by one.
+func (c *Core) FinalizeInFlightReqData(reqID string, statusCode int) {
+	if c.logRequestsInfo != "" {
+		c.LogCompletedRequests(reqID, statusCode)
+	}
+	c.inFlightReqData.InFlightReqMap.Delete(reqID)
+	c.inFlightReqData.InFlightReqCount.Dec()
+}
+
+// LoadInFlightReqData creates a snapshot map of the current
+// in-flight requests
+func (c *Core) LoadInFlightReqData() map[string]*InFlightReqData {
+	currentInFlightReqMap := make(map[string]*InFlightReqData)
+	c.inFlightReqData.InFlightReqMap.Range(func(key, value interface{}) bool {
+		// there is only one writer to this map, so skip checking for errors
+		v, _ := value.(*InFlightReqData)
+		currentInFlightReqMap[key.(string)] = v
+
+		return true
+	})
+
+	return currentInFlightReqMap
+}
+
+// UpdateInFlightReqData updates the data for a specific reqID with
+// the clientID
+func (c *Core) UpdateInFlightReqData(reqID, clientID string) {
+	v, ok := c.inFlightReqData.InFlightReqMap.Load(reqID)
+	if !ok {
+		c.Logger().Trace("failed to retrieve request with ID %v", reqID)
+		return
+	}
+	reqData, _ := v.(*InFlightReqData)
+	reqData.ClientID = clientID
+	c.inFlightReqData.InFlightReqMap.Store(reqID, reqData)
+}
+
+func (c *Core) logRequests(msg string, args ...interface{}) {
+	switch c.logRequestsInfo {
+	case "error":
+		c.Logger().Error(msg, args...)
+	case "warn":
+		c.Logger().Warn(msg, args...)
+	case "info":
+		c.Logger().Info(msg, args...)
+	case "debug":
+		c.Logger().Debug(msg, args...)
+	case "trace":
+		c.Logger().Trace(msg, args...)
+	}
+}
+
+// LogCompletedRequests Logs the completed request to the server logs
+func (c *Core) LogCompletedRequests(reqID string, statusCode int) {
+	v, ok := c.inFlightReqData.InFlightReqMap.Load(reqID)
+	if !ok {
+		c.logRequests(fmt.Sprintf("failed to retrieve request with ID %v", reqID))
+		return
+	}
+	// there is only one writer to this map, so skip checking for errors
+	reqData, _ := v.(*InFlightReqData)
+	c.logRequests("completed_request","client_id", reqData.ClientID, "client_address", reqData.ClientRemoteAddr, "status_code", statusCode, "request_path", reqData.ReqPath, "request_method", reqData.Method)
+}
+
+func (c *Core) ReloadLogRequestsInfo(){
+	conf := c.rawConfig.Load()
+	if conf == nil {
+		return
+	}
+	infoLevel := conf.(*server.Config).LogRequestsInfo
+	switch {
+	case log.LevelFromString(infoLevel) > 0:
+		c.logRequestsInfo = infoLevel
```

Comment:
This needs synchronization to avoid a race.

Reply:
I think you're underestimating what can happen with a data race.  It's not just a matter of "oh, we might see an incorrect value".  As far as I understand it, concurrent access without synchronization produces *undefined behaviour*, including memory corruption.  Depending on the platform, the reader could see a partial write, e.g. only half of a longword pointer could be written when the read is done, resulting in an invalid address and a panic.  These are rare and unlikely outcomes, but they could happen.  So data races should be treated as potential panics as far as I'm concerned.

I could also see us writing a test for the reload behaviour, and then we'd have the potential for the race tests failing due to the race, which would also be bad.  So yes, please add a lock or make logRequestsInfo an atomic.